### ISSUE TITLE
Bouncing ball with LazyTextBookDiscretePost

### ISSUE THREAD

**Issue by mforets** [JuliaReach/Reachability.jl#308]
For the given options in `test/Reachability/unit_solve_hybrid_bouncing_ball.jl`, and `mforets/bball_cleanup`, i get

```julia
# default algorithm
julia> @time sol = solve(system, options);
0.193210 seconds (151.92 k allocations: 7.185 MiB, 20.86% gc time)

julia> @time sol = solve(system, options, Reachability.BFFPSV18(),
                  Reachability.ReachSets.LazyTextbookDiscretePost());
38.213728 seconds (674.57 M allocations: 16.088 GiB, 18.12% gc time)
```

The purpose of this issue is to investigate this slowdown.

**Comment by mforets** [comment]
The problem is unrelated to the continuous post check,

```julia
        @time begin
        reach_tube = solve(ContinuousSystem(loc.A, X0.X, loc.U),
                           options_copy,
                           op=opC)
        end
```

prints `0.022208 seconds (16.90 k allocations: 1.209 MiB)` exactly two times for each scenario.

**Comment by mforets** [comment]
The difference is in the discrete post

```julia
        @time begin
        post(opD, HS, waiting_list, passed_list, loc_id, tube⋂inv, jumps,
             options)
        end
```

gives ` 0.063260 seconds (65.58 k allocations: 2.555 MiB)` and `40.283772 seconds (673.25 M allocations: 16.053 GiB, 18.91% gc time)` respectively.

**Comment by mforets** [comment]
Some inspection led to nailing down the slowdown to the operation

```julia
res = overapproximate(A⌜R⋂G⌟⋂I, op.options[:overapproximation];
                                  upper_bound=true)
```

Saving the object `A⌜R⋂G⌟⋂I` generated at runtime with [JLD.jl](https://github.com/JuliaIO/JLD.jl) then working with it separately led me to this performance issue:

> ```julia
julia> X = BallInf(ones(2), 1.0);
julia> H = LazySets.HalfSpace([-1.0, 0.0], 0.0);

julia> @time ρ(d, X∩H, upper_bound=true)
entering rho method
  0.000622 seconds (650 allocations: 51.813 KiB)

julia> @time ρ(d, X∩(X∩H), upper_bound=true)
entering rho method
...  # hundred of times!
entering rho method
  1.789980 seconds (1.02 M allocations: 63.388 MiB, 3.47% gc time)
```

One can expect in the second case an explosion in the number of calls to `ρ(,::Intersection)` because `line_search` is an iterative algorithm. There are three lazy intersections involved in `A⌜R⋂G⌟⋂I`.

See [LazySets #776](https://github.com/JuliaReach/LazySets.jl/issues/776).

**Comment by schillic** [comment]
concrete, `HPolytope`:
![concrete_hpolytope](https://user-images.githubusercontent.com/9656686/46918202-992f6480-cfcf-11e8-98f6-23ad41ffec7f.png)

lazy, `HPolytope`:
![lazy_hpolytope](https://user-images.githubusercontent.com/9656686/46918204-9d5b8200-cfcf-11e8-9a41-1da30d6e3290.png)

overapprox, `HPolytope`:
![overapprox_hpolytope](https://user-images.githubusercontent.com/9656686/46918205-a0567280-cfcf-11e8-84be-e7e70acd3e1a.png)

concrete, `HPolyhedron`:
![concrete_hpolyhedron](https://user-images.githubusercontent.com/9656686/46918207-a3e9f980-cfcf-11e8-962f-cfe4eaecfb0b.png)

lazy, `HPolyhedron`:
![lazy_hpolyhedron](https://user-images.githubusercontent.com/9656686/46918209-a64c5380-cfcf-11e8-9333-3d7e1bc9580e.png)

overapprox, `HPolyhedron`:
![overapprox_hpolyhedron](https://user-images.githubusercontent.com/9656686/46918210-a8aead80-cfcf-11e8-80c2-4d436d779ea5.png)

**Comment by mforets** [comment]
Fixed, see #326.
